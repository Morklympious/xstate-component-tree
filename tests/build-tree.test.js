"use strict";

const { Machine, interpret } = require("xstate");

const buildTree = require("../src/build-tree.js");

describe("xstate-component-tree", () => {
    it("should export a function by default", () => {
        expect(typeof buildTree).toBe("function");
    });

    it("should read a value from meta.component", async () => {
        const m = Machine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        component : "one",
                    },
                },
            },
        });

        const i = interpret(m);

        const out = await buildTree(i);

        expect(out).toMatchSnapshot();
    });

    it("should read values from meta.component", async () => {
        const m = Machine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        component : "one",
                    },

                    initial : "child",

                    states : {
                        child : {
                            meta : {
                                component : "child",
                            },
                        },
                    },
                },
            },
        });

        const i = interpret(m);

        const out = await buildTree(i);

        expect(out).toMatchSnapshot();
    });

    it("should read a value from meta.load", async () => {
        const m = Machine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        load : () => "one",
                    },
                },
            },
        });

        const i = interpret(m);

        const out = await buildTree(i);

        expect(out).toMatchSnapshot();
    });

    it("should read values from meta.component", async () => {
        const m = Machine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        load : () => "one",
                    },

                    initial : "child",

                    states : {
                        child : {
                            meta : {
                                load : () => "child",
                            },
                        },
                    },
                },
            },
        });

        const i = interpret(m);

        const out = await buildTree(i);

        expect(out).toMatchSnapshot();
    });

    it("should wait for all load() functions to resolve", async () => {
        const m = Machine({
            initial : "one",

            states : {
                one : {
                    meta : {
                        load : () => "one",
                    },

                    initial : "child",

                    states : {
                        child : {
                            meta : {
                                load : () => "child",
                            },
                        },
                    },
                },
            },
        });

        const i = interpret(m);

        const out = await buildTree(i);

        expect(out).toMatchSnapshot();
    });
});
