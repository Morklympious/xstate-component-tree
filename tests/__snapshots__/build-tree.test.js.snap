// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xstate-component-tree should read a value from meta.component 1`] = `
Map {
  "one" => Array [
    Object {
      "child": "one",
      "props": Object {},
    },
  ],
}
`;

exports[`xstate-component-tree should read a value from meta.load 1`] = `
Map {
  "one" => Array [
    Object {
      "child": "one",
      "props": Object {},
    },
  ],
}
`;

exports[`xstate-component-tree should read values from meta.component 1`] = `
Map {
  "one" => Array [
    Object {
      "child": "one",
      "props": Object {},
    },
    Object {
      "child": "child",
      "props": Object {},
    },
  ],
}
`;

exports[`xstate-component-tree should read values from meta.component 2`] = `
Map {
  "one" => Array [
    Object {
      "child": "one",
      "props": Object {},
    },
    Object {
      "child": "child",
      "props": Object {},
    },
  ],
}
`;

exports[`xstate-component-tree should wait for all load() functions to resolve 1`] = `
Map {
  "one" => Array [
    Object {
      "child": "one",
      "props": Object {},
    },
    Object {
      "child": "child",
      "props": Object {},
    },
  ],
}
`;
